File: src/ssh_manager/services/query.py
"""Read-only views over the manifest + inventory (`list` / `view`).

Returns *structured data* only; rendering lives in ``ssh_manager.render`` (rich), so
the surfaces (CLI/TUI/desktop) format it however they like. Provider *category*
powers ``--type``; free-form ``tags`` power ``--tag``.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

from ..core.inventory import Inventory, KeyRecord
from ..core.manifest import Host, Manifest
from ..providers.registry import resolve as resolve_provider
from ..util.errors import SshManagerError

NO_KEY = "no-key"
NEEDS_REDEPLOY = "needs-redeploy"
DEPLOYED = "deployed"


@dataclass(frozen=True)
class HostRow:
    alias: str
    hostname: str
    provider_label: str       # e.g. "github/vcs" or "server"
    key_name: str
    status: str               # no-key | needs-redeploy | deployed
    tags: list[str] = field(default_factory=list)


@dataclass(frozen=True)
class ProfileGroup:
    name: str
    empty: bool
    rows: list[HostRow] = field(default_factory=list)


@dataclass(frozen=True)
class DeploymentRow:
    target: str
    method: str
    verified: bool


@dataclass(frozen=True)
class HostDetail:
    profile: str
    alias: str
    hostname: str
    user: str
    port: int
    identity_file: str
    known_hosts: str
    provider_label: str
    key_name: str
    status: str
    fingerprint: str | None = None
    expires_on: str | None = None
    tags: list[str] = field(default_factory=list)
    raw_options: dict[str, str] = field(default_factory=dict)
    deployments: list[DeploymentRow] = field(default_factory=list)
    requires_vpn: bool = False
    vpn_name: str | None = None
    vpn_url: str | None = None


@dataclass(frozen=True)
class ProfileSummary:
    name: str
    key_scope: str
    rows: list[HostRow] = field(default_factory=list)


class Query:
    def __init__(self, manifest: Manifest, inventory: Inventory,
                 ssh_dir: Path, providers_file: Path) -> None:
        self._m = manifest
        self._inv = inventory
        self._ssh = ssh_dir
        self._categories = _load_categories(providers_file)
        self._by_path: dict[str, KeyRecord] = {r.path: r for r in inventory.keys.values()}

    def category_of(self, provider: str | None) -> str:
        if provider and provider in self._categories:
            return self._categories[provider]
        return resolve_provider(provider).category

    def groups(self, *, profile: str | None = None, provider: str | None = None,
               type_: str | None = None, tag: str | None = None) -> list[ProfileGroup]:
        filtered = any((profile, provider, type_, tag))
        out: list[ProfileGroup] = []
        for pname, prof in self._m.profiles.items():
            if profile and pname != profile:
                continue
            rows: list[HostRow] = []
            for host in prof.hosts:
                if provider and host.provider != provider:
                    continue
                if type_ and self.category_of(host.provider) != type_:
                    continue
                if tag and tag not in host.tags:
                    continue
                rows.append(self._row(pname, host))
            if rows or (not filtered and not prof.hosts):
                out.append(ProfileGroup(name=pname, empty=not prof.hosts, rows=rows))
        return out

    def detail(self, selector: str) -> ProfileSummary | HostDetail:
        if selector in self._m.profiles:
            prof = self._m.profiles[selector]
            return ProfileSummary(
                name=selector, key_scope=prof.key_scope,
                rows=[self._row(selector, h) for h in prof.hosts],
            )
        for pname, prof in self._m.profiles.items():
            for host in prof.hosts:
                if host.alias == selector:
                    return self._host_detail(pname, host)
        raise SshManagerError(f"no profile or host alias matches {selector!r}")

    # helpers
    def _provider_label(self, host: Host) -> str:
        cat = self.category_of(host.provider)
        return f"{host.provider}/{cat}" if host.provider else cat

    def _row(self, pname: str, host: Host) -> HostRow:
        kname = self._m.resolved_key_name(pname, host)
        rec = self._by_path.get(self._m.identity_file(pname, kname))
        return HostRow(
            alias=host.alias, hostname=host.hostname,
            provider_label=self._provider_label(host), key_name=kname,
            status=self._status(rec), tags=list(host.tags),
        )

    def _host_detail(self, pname: str, host: Host) -> HostDetail:
        kname = self._m.resolved_key_name(pname, host)
        ident = self._m.identity_file(pname, kname)
        rec = self._by_path.get(ident)
        fp = next((f for f, r in self._inv.keys.items() if r.path == ident), None)
        deps = [DeploymentRow(d.target, d.method, d.verified)
                for d in (rec.deployments if rec else [])]
        return HostDetail(
            profile=pname, alias=host.alias, hostname=host.hostname, user=host.user,
            port=host.port, identity_file=ident,
            known_hosts=self._m.known_hosts_file(pname),
            provider_label=self._provider_label(host),
            key_name=kname, status=self._status(rec),
            fingerprint=fp, expires_on=rec.expires_on if rec else None,
            tags=list(host.tags), raw_options=dict(host.raw_options), deployments=deps,
            requires_vpn=host.requires_vpn, vpn_name=host.vpn_name, vpn_url=host.vpn_url,
        )

    def _status(self, rec: KeyRecord | None) -> str:
        if rec is None:
            return NO_KEY
        return NEEDS_REDEPLOY if rec.needs_redeploy else DEPLOYED


def _load_categories(providers_file: Path) -> dict[str, str]:
    """Provider name -> category, from the layered catalog (built-ins + the user's
    providers.json if present, else the shipped default). Routing through the
    registry means `list --type`/category lookups honor the same fallback as
    deploy/resolve - not just a user file that may not exist."""
    from ..providers.registry import all_specs
    return {name: spec.category for name, spec in all_specs(providers_file).items()}

File: internal/services/query/query.go
// Package query is the read-only view layer over the manifest + inventory that
// backs list/view (and feeds audit), ported from src/ssh_manager/services/query.py.
// It returns structured data only; rendering lives with the caller, so the CLI and
// the future TUI format it however they like. Provider category powers --type; the
// host's free-form tags power --tag.
package query

import (
	"fmt"

	"github.com/simtabi/ssh-manager/internal/core/inventory"
	"github.com/simtabi/ssh-manager/internal/core/manifest"
	"github.com/simtabi/ssh-manager/internal/core/providers"
)

// Deployment status values.
const (
	NoKey         = "no-key"
	NeedsRedeploy = "needs-redeploy"
	Deployed      = "deployed"
)

// HostRow is one host in a list view.
type HostRow struct {
	Alias         string
	Hostname      string
	ProviderLabel string // e.g. "github/vcs" or "server"
	KeyName       string
	Status        string // no-key | needs-redeploy | deployed
	Tags          []string
}

// ProfileGroup is a profile and its (possibly filtered) host rows.
type ProfileGroup struct {
	Name  string
	Empty bool // the profile has no hosts at all
	Rows  []HostRow
}

// DeploymentRow is one recorded deployment of a key.
type DeploymentRow struct {
	Target   string
	Method   string
	Verified bool
}

// HostDetail is the resolved view of a single host (view <alias>).
type HostDetail struct {
	Profile       string
	Alias         string
	Hostname      string
	User          string
	Port          int
	IdentityFile  string
	KnownHosts    string
	ProviderLabel string
	KeyName       string
	Status        string
	Fingerprint   *string
	ExpiresOn     *string
	Tags          []string
	RawOptions    manifest.OrderedOptions
	Deployments   []DeploymentRow
	RequiresVPN   bool
	VPNName       *string
	VPNURL        *string
}

// ProfileSummary is the resolved view of a whole profile (view <profile>).
type ProfileSummary struct {
	Name     string
	KeyScope string
	Rows     []HostRow
}

// Query answers read-only questions about the manifest + inventory.
type Query struct {
	m          *manifest.Manifest
	inv        *inventory.Inventory
	categories map[string]string // provider name -> category (resolved once)
	byPath     map[string]inventory.KeyRecord
}

// New builds a Query. providersFile is the user's providers.json (may be absent;
// the embedded default catalog is used as the fallback).
func New(m *manifest.Manifest, inv *inventory.Inventory, providersFile string) *Query {
	cats := map[string]string{}
	for name, spec := range providers.AllSpecs(providersFile) {
		cats[name] = spec.Category
	}
	byPath := make(map[string]inventory.KeyRecord, len(inv.Keys))
	for _, r := range inv.Keys {
		byPath[r.Path] = r
	}
	return &Query{m: m, inv: inv, categories: cats, byPath: byPath}
}

// categoryOf mirrors query.category_of: the catalog category if known, else
// "server" (the generic-ssh fallback from registry.resolve).
func (q *Query) categoryOf(provider string) string {
	if provider != "" {
		if c, ok := q.categories[provider]; ok {
			return c
		}
	}
	return "server"
}

// Groups returns the list view, filtered by any of profile/provider/type_/tag
// (empty string means no filter on that axis).
func (q *Query) Groups(profile, provider, type_, tag string) ([]ProfileGroup, error) {
	filtered := profile != "" || provider != "" || type_ != "" || tag != ""
	var out []ProfileGroup
	for _, pname := range q.m.ProfileNames() {
		if profile != "" && pname != profile {
			continue
		}
		prof := q.m.Profiles[pname]
		var rows []HostRow
		for _, host := range prof.Hosts {
			hp := deref(host.Provider)
			if provider != "" && hp != provider {
				continue
			}
			if type_ != "" && q.categoryOf(hp) != type_ {
				continue
			}
			if tag != "" && !contains(host.Tags, tag) {
				continue
			}
			row, err := q.row(pname, host)
			if err != nil {
				return nil, err
			}
			rows = append(rows, row)
		}
		if len(rows) > 0 || (!filtered && len(prof.Hosts) == 0) {
			out = append(out, ProfileGroup{Name: pname, Empty: len(prof.Hosts) == 0, Rows: rows})
		}
	}
	return out, nil
}

// Detail returns a *ProfileSummary when selector names a profile, else a
// *HostDetail when it matches a host alias, else an error.
func (q *Query) Detail(selector string) (any, error) {
	if prof, ok := q.m.Profiles[selector]; ok {
		var rows []HostRow
		for _, h := range prof.Hosts {
			row, err := q.row(selector, h)
			if err != nil {
				return nil, err
			}
			rows = append(rows, row)
		}
		return &ProfileSummary{Name: selector, KeyScope: prof.KeyScope, Rows: rows}, nil
	}
	for _, pname := range q.m.ProfileNames() {
		for _, host := range q.m.Profiles[pname].Hosts {
			if host.Alias == selector {
				return q.hostDetail(pname, host)
			}
		}
	}
	return nil, fmt.Errorf("no profile or host alias matches %q", selector)
}

func (q *Query) providerLabel(host manifest.Host) string {
	provider := deref(host.Provider)
	cat := q.categoryOf(provider)
	if provider != "" {
		return provider + "/" + cat
	}
	return cat
}

func (q *Query) row(pname string, host manifest.Host) (HostRow, error) {
	kname, err := q.m.ResolvedKeyName(pname, host)
	if err != nil {
		return HostRow{}, err
	}
	rec, ok := q.byPath[q.m.IdentityFile(pname, kname)]
	return HostRow{
		Alias: host.Alias, Hostname: host.Hostname,
		ProviderLabel: q.providerLabel(host), KeyName: kname,
		Status: status(rec, ok), Tags: tagsOf(host),
	}, nil
}

func (q *Query) hostDetail(pname string, host manifest.Host) (*HostDetail, error) {
	kname, err := q.m.ResolvedKeyName(pname, host)
	if err != nil {
		return nil, err
	}
	ident := q.m.IdentityFile(pname, kname)
	rec, ok := q.byPath[ident]
	var fp *string
	for f, r := range q.inv.Keys {
		if r.Path == ident {
			fpCopy := f
			fp = &fpCopy
			break
		}
	}
	var deps []DeploymentRow
	var expiresOn *string
	if ok {
		for _, d := range rec.Deployments {
			deps = append(deps, DeploymentRow{Target: d.Target, Method: d.Method, Verified: d.Verified})
		}
		expiresOn = rec.ExpiresOn
	}
	return &HostDetail{
		Profile: pname, Alias: host.Alias, Hostname: host.Hostname, User: host.User,
		Port: host.Port, IdentityFile: ident, KnownHosts: q.m.KnownHostsFile(pname),
		ProviderLabel: q.providerLabel(host), KeyName: kname, Status: status(rec, ok),
		Fingerprint: fp, ExpiresOn: expiresOn, Tags: tagsOf(host), RawOptions: host.RawOptions,
		Deployments: deps, RequiresVPN: host.RequiresVPN, VPNName: host.VPNName, VPNURL: host.VPNURL,
	}, nil
}

func status(rec inventory.KeyRecord, found bool) string {
	if !found {
		return NoKey
	}
	if rec.NeedsRedeploy() {
		return NeedsRedeploy
	}
	return Deployed
}

func deref(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

func tagsOf(h manifest.Host) []string {
	out := make([]string, len(h.Tags))
	copy(out, h.Tags)
	return out
}

func contains(xs []string, v string) bool {
	for _, x := range xs {
		if x == v {
			return true
		}
	}
	return false
}
